// coc-tsserver/lib/server/features/documentHighlight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
const typeConverters = tslib_1.__importStar(require("../utils/typeConverters"));
const arrays_1 = require("../../utils/arrays");
class TypeScriptDocumentHighlightProvider {
    constructor(client) {
        this.client = client;
    }
    provideDocumentHighlights(resource, position, token) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const file = this.client.toPath(resource.uri);
            if (!file)
                return [];
            const args = Object.assign({}, typeConverters.Position.toFileLocationRequestArgs(file, position), { filesToSearch: [file] });
            try {
                const response = yield this.client.execute('documentHighlights', args, token);
                if (response.type !== 'response' || !response.body) {
                    return [];
                }
                return arrays_1.flatten(response.body
                    .filter(highlight => highlight.file === file)
                    .map(convertDocumentHighlight));
            }
            catch (_e) {
                return [];
            }
        });
    }
}
exports.default = TypeScriptDocumentHighlightProvider;
function convertDocumentHighlight(highlight) {
    return highlight.highlightSpans.map(span => {
        return {
            range: typeConverters.Range.fromTextSpan(span),
            kind: span.kind === 'writtenReference' ? vscode_languageserver_protocol_1.DocumentHighlightKind.Write : vscode_languageserver_protocol_1.DocumentHighlightKind.Read
        };
    });
}
