// coc-yank/lib/db.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const util_1 = require("./util");
const uuid = require("uuid/v1");
class DB {
    constructor(directory, maxsize) {
        this.maxsize = maxsize;
        this.file = path_1.default.join(directory, 'yank.json');
    }
    load() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let stat = yield util_1.statAsync(this.file);
            if (!stat || !stat.isFile())
                return [];
            let content = yield util_1.readFile(this.file);
            return JSON.parse(content);
        });
    }
    add(content, regtype, path, filetype) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let items = yield this.load();
            if (items.length == this.maxsize) {
                items.pop();
            }
            items.unshift({ id: uuid(), content, regtype, path, filetype });
            yield util_1.writeFile(this.file, JSON.stringify(items, null, 2));
        });
    }
    delete(uid) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let items = yield this.load();
            let idx = items.findIndex(o => o.id == uid);
            if (idx !== -1) {
                items.splice(idx, 1);
                yield util_1.writeFile(this.file, JSON.stringify(items));
            }
        });
    }
}
exports.default = DB;
